File: src/main/java/me/melijn/jda/commands/HelpCommand.java
package me.melijn.jda.commands;

import me.melijn.jda.blub.Category;
import me.melijn.jda.blub.Command;
import me.melijn.jda.blub.CommandEvent;
import me.melijn.jda.utils.Embedder;

import java.util.List;
import java.util.stream.Collectors;

import static me.melijn.jda.Melijn.PREFIX;

public class HelpCommand extends Command {

    public HelpCommand() {
        this.commandName = "help";
        this.description = "Shows you help for commands";
        this.usage = PREFIX + commandName + " [list | command]";
        this.aliases = new String[]{"commands", "cmds", "cmd"};
        this.category = Category.DEFAULT;
        this.id = 56;
    }

    @Override
    protected void execute(CommandEvent event) {
        String[] args = event.getArgs().split("\\s+");
        if (event.getArgs().isEmpty()) {
            if (event.getGuild() == null) {
                event.reply("https://melijn.com/commands or >help list");
            } else {
                event.reply("https://melijn.com/server/" + event.getGuildId() + "/commands or " + event.getVariables().prefixes.get(event.getGuildId()) + "help list");
            }
        } else {
            if (args[0].equalsIgnoreCase("list")) {
                List<Command> commandList = event.getClient().getCommands();

                Embedder eb = new Embedder(event.getVariables(), event.getGuild());
                eb.setTitle("Command List", "https://melijn.com/commands");

                String DEFAULT = commandList.stream()
                        .filter(command -> command.getCategory() == Category.DEFAULT)
                        .map(Command::getCommandName)
                        .collect(Collectors.joining("\n"));

                String MUSIC = commandList.stream().
                        filter(command -> command.getCategory() == Category.MUSIC)
                        .map(Command::getCommandName)
                        .collect(Collectors.joining("\n"));

                String UTIL = commandList.stream()
                        .filter(command -> command.getCategory() == Category.UTILS)
                        .map(Command::getCommandName)
                        .collect(Collectors.joining("\n"));

                String MANAGEMENT = commandList.stream()
                        .filter(command -> command.getCategory() == Category.MANAGEMENT)
                        .map(Command::getCommandName)
                        .collect(Collectors.joining("\n"));

                String FUN = commandList.stream()
                        .filter(command -> command.getCategory() == Category.FUN)
                        .map(Command::getCommandName)
                        .collect(Collectors.joining("\n"));

                //Default commands
                eb.addField("Default", DEFAULT, true);

                //Music commands
                eb.addField("Music", MUSIC, true);

                //Util commands
                eb.addField("Utilities", UTIL, true);

                //Management commands
                eb.addField("Management", MANAGEMENT, true);

                //Fun commands
                eb.addField("Fun", FUN, true);

                eb.setFooter("Total command: " + commandList.size(), null);

                event.reply(eb.build());
                return;
            }

            for (Command command : event.getClient().getCommands()) {
                if (command.isCommandFor(args[0])) {
                    event.reply("" +
                            "Help off **" + command.getCommandName() + "**\n" +
                            "**Usage:**  `" + command.getUsage() + "`\n" +
                            (command.getAliases().length > 0 ? ("**Aliases:** `" + String.join(", ", command.getAliases()) + "`\n") : "") +
                            "**Description:**  " + command.getDescription() +
                            (command.getExtra().isEmpty() ? "" : "\n**Extra:**  " + command.getExtra()));
                    return;
                }
            }
        }

    }
}

File: src/main/java/me/melijn/melijnbot/commands/utility/HelpCommand.kt
package me.melijn.melijnbot.commands.utility

import me.melijn.melijnbot.objects.command.AbstractCommand
import me.melijn.melijnbot.objects.command.CommandCategory
import me.melijn.melijnbot.objects.command.CommandContext
import me.melijn.melijnbot.objects.embed.Embedder
import me.melijn.melijnbot.objects.translation.i18n
import me.melijn.melijnbot.objects.utils.sendEmbed
import me.melijn.melijnbot.objects.utils.sendMsg
import me.melijn.melijnbot.objects.utils.sendSyntax
import java.util.stream.Collectors


class HelpCommand : AbstractCommand("command.help") {

    init {
        id = 6
        name = "help"
        aliases = arrayOf("commands", "command", "cmds", "cmd")
        commandCategory = CommandCategory.UTILITY
    }

    override suspend fun execute(context: CommandContext) {
        val args = context.args
        val language = context.getLanguage()
        if (args.isEmpty()) {
            val part = if (context.isFromGuild) "server" else "pm"
            val response = i18n.getTranslation(language, "$root.response1.$part")
            val msg = replaceArgs(response, context.getGuildId(), context.usedPrefix)
            sendMsg(context, msg)
            return
        }

        val commandList = context.getCommands()
        if (args[0].equals("list", ignoreCase = true)) {

            val title = i18n.getTranslation(language, "$root.response2.title")
            val util = i18n.getTranslation(language, "$root.response2.field1.title")
            val administration = i18n.getTranslation(language, "$root.response2.field2.title")
            val moderation = i18n.getTranslation(language, "$root.response2.field3.title")
            val music = i18n.getTranslation(language, "$root.response2.field4.title")
            val image =i18n.getTranslation(language, "$root.response2.field5.title")
            val animal = i18n.getTranslation(language, "$root.response2.field6.title")
            val anime = i18n.getTranslation(language, "$root.response2.field7.title")
            val supporter = i18n.getTranslation(language, "$root.response2.field8.title")
            val commandAmount = i18n.getTranslation(language, "$root.response2.footer")
                    .replace("%cmdCount%", commandList.size.toString())

            val eb = Embedder(context)
            eb.setTitle(title, "https://melijn.com/commands")

            eb.addField(util, commandListString(commandList, CommandCategory.UTILITY), true)
            eb.addField(administration, commandListString(commandList, CommandCategory.ADMINISTRATION), true)
            eb.addField(moderation, commandListString(commandList, CommandCategory.MODERATION), true)
            eb.addField(music, commandListString(commandList, CommandCategory.MUSIC), true)
            eb.addField(image, commandListString(commandList, CommandCategory.IMAGE), true)
            eb.addField(animal, commandListString(commandList, CommandCategory.ANIMAL), true)
            eb.addField(anime, commandListString(commandList, CommandCategory.ANIME), true)
            eb.addField(supporter, commandListString(commandList, CommandCategory.SUPPORTER), true)

            eb.setFooter(commandAmount, null)

            sendEmbed(context, eb.build())
            return
        }

        for (command in commandList) {
            if (!command.isCommandFor(args[0])) continue
            var msg = i18n.getTranslation(language, "$root.response3.line1")
            msg += i18n.getTranslation(language, "$root.response3.line2")
            if (command.aliases.isNotEmpty()){
                msg += i18n.getTranslation(language, "$root.response3.line3")
            }
            msg += i18n.getTranslation(language, "$root.response3.line4")
            if (command.help != "empty") {
                msg += i18n.getTranslation(language, "$root.response3.line5")
            }
            msg += i18n.getTranslation(language, "$root.response3.line6")

            msg = replaceCmdVars(msg, context, command)

            sendMsg(context, msg)
            return
        }

        sendSyntax(context, syntax)
    }


    private suspend fun replaceCmdVars(msg: String, context: CommandContext, command: AbstractCommand): String = msg
            .replace("%cmdName%", command.name)
            .replace("%cmdSyntax%", i18n.getTranslation(context.getLanguage(), command.syntax))
            .replace("%cmdAliases%", command.aliases.joinToString())
            .replace("%cmdDescription%", i18n.getTranslation(context.getLanguage(), command.description))
            .replace("%cmdHelp%", i18n.getTranslation(context.getLanguage(), command.help))
            .replace("%cmdCategory%", command.commandCategory.toString())
            .replace("%prefix%", context.usedPrefix)


    private fun commandListString(list: Set<AbstractCommand>, category: CommandCategory): String = list
            .stream()
            .filter { command -> command.commandCategory == category }
            .map { fCommand -> fCommand.name }
            .collect(Collectors.joining("\n"))


    private fun replaceArgs(string: String, guildId: Long, usedPrefix: String): String = string
            .replace("%guildId%", guildId.toString())
            .replace("%prefix%", usedPrefix)

}